src/servicehandler.cpp

```diff
@@ -194,122 +92,171 @@ ServiceHandler::ServiceHandler()
-        return false;
-
-    ServiceInterface *service = d->getServicePlugin(notebook, storage);
+    ServiceInterface *plugin = service(serviceId);
+    if (!plugin) {
+        plugin = service(defaultName);
+    }
 
-    if (service) {
-        bool res = service->multiCalendar();
-        if (!res) {
-            d->mError = (ServiceHandler::ErrorCode) service->error(); //Right now convert directly
-        }
+    d->mError = ErrorOk;
+    if (plugin) {
+        bool res = plugin->multiCalendar();
+        d->mError = (ServiceHandler::ErrorCode) plugin->error(); //Right now convert directly
         return res;
     } else {
         return false;
     }
 }
 
-QString ServiceHandler::emailAddress(const Notebook::Ptr &notebook, const ExtendedStorage::Ptr &storage)
+QString ServiceHandler::emailAddress(const Notebook::Ptr &notebook)
 {
-    if (storage.isNull() || notebook.isNull())
-        return QString();
-
-    ServiceInterface *service = d->getServicePlugin(notebook, storage);
+    ServiceInterface *plugin = service(notebook->pluginName());
+    if (!plugin) {
+        plugin = service(defaultName);
+    }
 
-    if (service) {
-        QString res = service->emailAddress(notebook);
+    d->mError = ErrorOk;
+    if (plugin) {
+        QString res = plugin->emailAddress(notebook);
         if (res.isNull()) {
-            d->mError = (ServiceHandler::ErrorCode) service->error(); //Right now convert directly
+            d->mError = (ServiceHandler::ErrorCode) plugin->error(); //Right now convert directly
         }
         return res;
     } else {
         return QString();
     }
 }
 
-QString ServiceHandler::displayName(const Notebook::Ptr &notebook, const ExtendedStorage::Ptr &storage)
+QString ServiceHandler::displayName(const Notebook::Ptr &notebook)
 {
-    if (storage.isNull() || notebook.isNull())
-        return QString();
-
-    ServiceInterface *service = d->getServicePlugin(notebook, storage);
+    ServiceInterface *plugin = service(notebook->pluginName());
+    if (!plugin) {
+        plugin = service(defaultName);
+    }
 
-    if (service) {
-        QString res = service->displayName(notebook);
+    d->mError = ErrorOk;
+    if (plugin) {
+        QString res = plugin->displayName(notebook);
         if (res.isNull()) {
-            d->mError = (ServiceHandler::ErrorCode) service->error(); //Right now convert directly
+            d->mError = (ServiceHandler::ErrorCode) plugin->error(); //Right now convert directly
         }
         return res;
     } else {
         return QString();
     }
 }
 
-int ServiceHandler::downloadAttachment(const Notebook::Ptr &notebook, const ExtendedStorage::Ptr &storage,
-                                       const QString &uri, const QString &path)
+int ServiceHandler::downloadAttachment(const Notebook::Ptr &notebook, const QString &uri, const QString &path)
 {
-    if (storage.isNull() || notebook.isNull())
+    if (notebook->isRunTimeOnly()) {
```

Comment:
Was already checking this, but wonder if needed. A runtime notebook could in theory do some downloading? At the moment not used, of course. Same thing on deletion side.

Reply:
It was there already in the `executePlugin()` routine. I removed this routine, but kept the chek (I've removed the readOnly() one considering that it's a hint only).

Do you prefer to remove this test ?

Reply:
Ah, no sorry. I mixed up with the sendResponse() case and friends. I've removed the test here and in delete.